Changelog 136 (#100)
intellij-rust/intellij-rust.github.io: _posts/2020-11-27-changelog-136.markdown

---
layout: post
title: "IntelliJ Rust Changelog #136"

> date: 2020-11-27 09:07:11 +0300

Reviewer: Don't forget to change date and file name to actual one

---

## New Features

<!-- https://github.com/intellij-rust/intellij-rust/pull/6378 -->
* Introduce new experimental engine for name resolution (disabled by default).
  The new algorithm (aka name resolution 2.0) uses the fixed-point iteration approach and aims at better support for

>   complex glob imports (`use foo::*`) and re-exports. Also, it will allow the plugin to properly expand the up-coming
>   macros 2.0. Note that the new algorithm works for top-level items only, so local variables, as well as type inference

Reviewer: ```suggestion
  complex glob imports (`use foo::*`) and re-exports. Also, it will allow the plugin to expand the up-coming [macros 2.0](https://rust-lang.github.io/rfcs/1584-macros.html) properly.
  The new algorithm works for top-level items only, so local variables, as well as type inference
```

>   functionality, won’t be affected. For now, the newer version of the algorithm is available as an experimental feature.

Reviewer: ```suggestion
  functionality, won’t be affected.
```

"experimental feature" is a concrete term in our IDEs as a result, it's not true since we introduce user-visible option to enable it. Let's just drop this sentence

>   If you are ready to give it a try, enable the corresponding checkbox in
>   `Settings / Preferences | Languages & Frameworks | Rust | Use experimental name resolution engine`
>   (note that the option is per-project). This feature is in early stages of development, so we encourage you to try

Reviewer: ```suggestion
  If you are ready to give it a try, enable `Use experimental name resolution engine` option in
  `Settings / Preferences | Languages & Frameworks | Rust` settings.
  The feature is in the early stages of development, so we encourage you to try
```

>    it out and share if you face any issues. Please `Help | Create New Issue` action in your IDE or our

Reviewer: ```suggestion
   it out and share if you face any issues. Please use `Help | Create New Issue` action in your IDE or our
```

   [issue tracker](https://github.com/intellij-rust/intellij-rust/issues/new) directly.

<!-- https://github.com/intellij-rust/intellij-rust/pull/6276 -->

> * Support `Go to Declaration` for Cargo [features](https://doc.rust-lang.org/cargo/reference/features.html) in

Reviewer: Let's add a shortcut of the action

>   `#[cfg()]` and `#[cfg_attr()]` attributes.

Reviewer: ```suggestion
  [`cfg`](https://doc.rust-lang.org/reference/conditional-compilation.html#the-cfg-attribute) and [`cfg_attr`](https://doc.rust-lang.org/reference/conditional-compilation.html#the-cfg_attr-attribute) attributes.
```

>   `#[cfg()]` and `#[cfg_attr()]` attributes.

Reviewer: Gif

<!-- https://github.com/intellij-rust/intellij-rust/pull/6396 -->
* Take into account macro invocations in type context, e.g. `const FOO: foobar!() = 1`

* Compiler errors detection improvements:
  <!-- https://github.com/intellij-rust/intellij-rust/pull/6413 -->
    * Detect attributes that require parentheses but have no parentheses, and suggest `Add parentheses`
      quick-fix (by [@zeroeightysix])

  <!-- https://github.com/intellij-rust/intellij-rust/pull/6372 -->
    * Detect the missing type of `const` and `static` items as a semantic error and produce a user-friendly error message.
      Also, provide a quick-fix to add explicit type in such cases (by [@Kobzol])

  <!-- https://github.com/intellij-rust/intellij-rust/pull/6352 -->

>     * Show error if `repr` attribute doesn't have parentheses and provide quick-fix to add them (by [@zeroeightysix])

Reviewer: Superseded by the first item in this group. Let's drop the current one and add a comment with https://github.com/intellij-rust/intellij-rust/pull/6352 to first item

    <!-- https://github.com/intellij-rust/intellij-rust/pull/6319 -->
    * Detect [E0057](https://doc.rust-lang.org/error-index.html#E0057) in some cases (by [@zeroeightysix])

## Performance Improvements

<!-- https://github.com/intellij-rust/intellij-rust/pull/6395 -->
* Slightly optimize highlighting and other analysis

## Fixes

<!-- https://github.com/intellij-rust/intellij-rust/pull/6380 -->

> * MACRO: fix macro group matching in some cases. In particular, this fixes completion in `tokio 0.3`

Reviewer: ```suggestion
* Fix macro group matching in some cases. In particular, this fixes completion in `tokio 0.3`
```

<!-- https://github.com/intellij-rust/intellij-rust/pull/6406 -->

> * Fix 2020.3 EAPs compatibility

Reviewer: Let's drop it since it's rather unclear
